Allow to use any storage policy for external aggregation/sorting

azat opened the pull request:
### Changelog category (leave one):
- Improvement

### Changelog entry (a [user-readable short description](https://github.com/ClickHouse/ClickHouse/blob/master/docs/changelog_entry_guidelines.md) of the changes that goes into CHANGELOG.md):
Allow to use any storage policy (i.e. object storage, such as S3) for external aggregation/sorting

Fixes: https://github.com/ClickHouse/ClickHouse/issues/42157

azat committed:
Allow to use any storage policy for external aggregation/sorting

v2: Use VFS layer for cleanup temporary disk at start
v3: Relax documentation for tmp_policy setting (any disk is supported)
v4: Use removeFile() over removeRecursive() while cleaning up temporary files
v5: Add a test for cleanup tmp disk at start

alexey-milovidov reviewed (approved):
Amazing!

The only potential downside is - garbage accumulation on unclean restarts. But it is tolerable.

vdimir reviewed (approved):
Great, thanks for this PR!

We indeed use only `disk` methods (previously, it was direct file operations when this restriction was added):
https://github.com/ClickHouse/ClickHouse/blob/2f36b294d22b17a7a56feec5a1e67aea048a527d/src/Interpreters/TemporaryDataOnDisk.cpp#L155-L158

The only thing I'm not sure about is whether `removeRecursive` is overkill here, if it adds too much overhead for remote disks, we might consider replacing it with `removeFileIfExists` later.

---

For now, we may want to update the doc:
https://github.com/ClickHouse/ClickHouse/blob/2f36b294d22b17a7a56feec5a1e67aea048a527d/src/Core/ServerSettings.cpp#L171

azat commented:
>For now, we may want to update the doc:

Thanks, fixed.

>The only thing I'm not sure about is whether removeRecursive is overkill here, if it adds too much overhead for remote disks, we might consider replacing it with removeFileIfExists later.

Replace with `removeFile`, it has a try/catch anyway. Thanks!

azat committed:
local: create data directory prio to setting up data for temporary storage

Previously this code relies on that fact that setupTmpPath() will create
the directory recursively, found it because
03321_clickhouse_local_initialization_not_too_slow_even_under_sanitizers
started to fail.
